MarkBind/markbind
# Wrap fenced code blocks on-demand

**What is the purpose of this pull request?**


Resolves #1289 



**Overview of changes:**
* Place "copy" buttons for code blocks inside a custom container (in order to provide room for more buttons with different functionalities).
* Add a "wrap text" button plugin for code blocks.

**Anything you'd like to highlight / discuss:**
* Modifications were made for `markbind.css`; thus, frontend bundles should be updated for the best results.
* In case the first line of the code block is long, the button will appear overlaid on the code like so:
![image](https://user-images.githubusercontent.com/69278589/133271095-91994371-0073-4683-9dd8-f7f6a0757900.png)

    * Note: the original implementation of the "copy" button also had this issue.
    * Thus, a change in the positioning/design of the buttons might be preferred. Here's an example:
       
     ![potential](https://user-images.githubusercontent.com/69278589/133269219-b561866b-8490-4c28-b424-de334ff47a0e.png)
Feel free to suggest more designs.

**Testing instructions:**
Since there doesn't seem to be any automated tests for the basis of this feature (i.e., the "copy" code block plugin), this feature also does not have such tests. If there are automated tests for the "copy" code block button that I might have missed, please let me know; otherwise, feel free to ask me to write automated tests for both plugins.

With that said, here are some steps to test this feature manually:
1. Initialize a new MarkBind project (if none exist).
2. Add `codeBlockWrapButtons` to the `plugins` array in the root's `site.json`.
    * If `site.json` does not have a `plugins` attribute, copy and paste the following before the last closing curly brace at the end of the file:
    ```json
    "plugins": [
      "codeBlockWrapButtons"
    ],
    ```
3. Add a code block to some page in the project (if none exist).
4. Write a long line of code inside the code block.
    * E.g., `This is a very very very very very very very very very very very very very very very very very very very very very very very very very very very very very long line of code`.
4. Serve the project using `markbind serve -d`.
5. The button should appear on the top-right corner of any code block like so:
6. Click the button to enable line wrapping. The result should look similar to this:
7. Click the button again to disable line wrapping. The code block should return to its original state.
* Note: other scenarios can also be tested similarly, such as using both `codeBlockCopyButtons` and `codeBlockWrapButtons` plugins or using the "wrap text" button with code block consisting of short lines.



**Proposed commit message: (wrap lines at 72 characters)**

```
Wrap fenced code blocks on-demand

Fenced code blocks are displayed without text wrapping and do not offer
a way of doing so.

In some cases, viewing a code snippet with text wrapping is preferred,
since it allows the readers to read the code without the need to scroll
sideways for long lines.

Let's add a button, in the form of a plugin, that allows the readers to
toggle text-wrapping for code blocks.

Implementing this button as a plugin can simplify the addition of this
feature in MarkBind websites due to its similarity with the "copy" code
block button feature.
```


---

**Checklist:** :ballot_box_with_check:



    * Some changes to the original `codeBlockCopyButton` code were made in order to allow for the implementation of more buttons with different functions.
